packages/flutter/lib/src/material/tabs.dart

```diff
@@ -432,63 +474,77 @@ class TabBarSelection {
     _performance
       ..progress = progress
       ..forward().then((_) {
-        if (onChanged != null)
-          onChanged();
+        if (config.onChanged != null)
+          config.onChanged(_index);
         _indexIsChanging = false;
       });
   }
 
   int get previousIndex => _previousIndex;
   int _previousIndex = 0;
+
+  Widget build(BuildContext context) {
+    return new _TabBarSelection(selection: this, child: config.child);
+  }
 }
 
-/// A tab strip, consisting of several TabLabels and a TabBarSelection.
-/// The TabBarSelection can be used to link this to a TabBarView.
+
+/// Displays a horizontal row of tabs, one per label. If isScrollable is
+/// true then each tab is as wide as needed for its label and the entire
+/// [TabBar] is scrollable. Otherwise each tab gets an equal share of the
+/// available space. A [TabBarSelection] widget ancestor must have been
+/// built to enable saving and monitoring the selected tab.
 ///
 /// Tabs must always have an ancestor Material object.
 class TabBar extends Scrollable {
   TabBar({
     Key key,
     this.labels,
-    this.selection,
     this.isScrollable: false
   }) : super(key: key, scrollDirection: ScrollDirection.horizontal) {
     assert(labels != null);
     assert(labels.length > 1);
-    assert(selection != null);
-    assert(selection.maxIndex == labels.length - 1);
   }
 
   final Iterable<TabLabel> labels;
-  final TabBarSelection selection;
   final bool isScrollable;
 
   _TabBarState createState() => new _TabBarState();
 }
 
 class _TabBarState extends ScrollableState<TabBar> {
+
+  TabBarSelectionState _selection;
+  bool _indexIsChanging = false;
+
+  int get _tabCount => config.labels.length;
+
+  void _addSelectionListeners(TabBarSelectionState selection) {
```

Comment:
Since TabBarSelection's descendants are adding listeners to its performance, perhaps they should also add a TBD deactivate() listener which would they'd use to remove the listeners they added? An alternate version of the same thing would be to have add/removePerformanceFooListener methods on TabBarSelection that DTRT on deactivate?